### PR TITLE
SCM: Support past commit message navigation

### LINKED ISSUE #26952: SCM: Support past commit messages navigation
- VSCode Version: 1.13.0-insider
- OS Version: any

This is a feature request. 

In the search boxes (both local and multifile) you can use ALT-↑ to navigate through your search history. I suggest the same is implemented for the commit message box in the git sidebar. This should be no cheap `git log` substitute - you should only see the (last X?) commit messages you have entered yourself on that machine, in the order in which you have entered them. Just a local history of what was entered in that box.

This is somewhat similar to https://github.com/Microsoft/vscode/issues/6190, but different - it has no relation to undoing a commit, it should be there in the commit message box even if there is nothing to commit.

Thanks

### PR DESCRIPTION
**Pull request by meganrogge**
This PR addresses this [feature request](https://github.com/microsoft/vscode/issues/26952)

### CODE REVIEW

**Comment by joaomoreno** [comment]
@meganrogge We add `verification-needed` on feature requests that don't end up on an endgame test plan. @mattbierner can clarify that a little bit since we're currently in endgame.

File: src/vs/workbench/contrib/scm/browser/scm.contribution.ts
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/browser/scm.contribution.ts:239]: Should this work like the search viewlet instead? Pressing up while on the first line should go back in history, etc. Do we need an `Alt` modifier?

File: src/vs/workbench/contrib/scm/common/scmService.ts
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/common/scmService.ts:74]: Using `:any` is a big no-no. What type is this?

File: src/vs/workbench/contrib/scm/common/scmService.ts
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/common/scmService.ts:101]: Why do we have this event at all? Seems like it's not connected to anything.

File: src/vs/workbench/contrib/scm/common/scmService.ts
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/common/scmService.ts:106]: Why do we have an extra `:latest` key here? Do we really need a separate key to implement this?

File: src/vs/workbench/contrib/scm/common/scmService.ts
Comment (joaomoreno) [reply src/vs/workbench/contrib/scm/common/scmService.ts:101]: Also, it's important to keep the code style: remember to place fields and properties together, before the constructor.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
At first glance, there is quite some work to do here.

The `SCMInput`'s API should really not change much, adding 3 methods to it seems like an overkill. Could we reuse the [`HistoryNavigator`](https://github.com/microsoft/vscode/blob/094d4bb3f33265d4a49c6898ba2fe907f7d1275b/src/vs/base/common/history.ts#L8-L8) here?

Also, that addition of `build/monaco/api.js` seems a mistake, let's remove it.

File: src/vs/workbench/contrib/scm/common/scmService.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/scm/common/scmService.ts:106]: To save the SCM input when it hasn't yet been entered (per this [issue](https://github.com/microsoft/vscode/issues/41232)). Input in the SCM isn't part of the commit history and shouldn't be until enter has been pressed. Alternatively, the latest typed input could be part of the history array at a pre-defined index.

File: src/vs/workbench/contrib/scm/browser/scm.contribution.ts
Comment (meganrogge) [reply src/vs/workbench/contrib/scm/browser/scm.contribution.ts:239]: I was planning to replace that with [this ](https://github.com/microsoft/vscode/issues/107458#issue-709102124)logic

**Comment by meganrogge** [comment]
@joaomoreno There are a few issues with the new approach:

1. If we only save the current value when the arrows are pressed, then on window reload, 
text typed but not yet committed will not be saved (unless we use the saveCurrent method).

2. I haven't found a way using the historyNavigator to replace / delete a value. I need this so that the current element can be replaced with a new value and so that there isn't a backlog of uncommitted messages. 

For example, let's say I have 3 messages successfully committed.

message 1 
message 2
message 3

and I can navigate between them. so i'm typing message 4 when I want to reference 
message 2. I go up to message 2, return to message 4, successfully commit it.

then what will be in the history at that time is:

message 1
message 2
message 3 
message 4 (before looking up at message 2)
message 4 completed

or I return to message 4 and change it completely to hello.

what i'll see in the history if I go back through it is:

message 1
message 2
message 3
message 4 
hello

even though message 4 was not a successful commit

I'm not sure how to solve this. what I'd want to see is:

message 1
message 2
message 3
hello

do you agree this would be the expectation?

also, on the first pass through it's doing a strange thing with the order. on the way back down it's resolved and appears to store them in the correct order when debugged.

**Comment by joaomoreno** [comment]
> If we only save the current value when the arrows are pressed, then on window reload,
text typed but not yet committed will not be saved (unless we use the saveCurrent method).

We'd save the value when (1) the inbox box changes in interesting ways (the `fromKeyboard` story), (2) when navigating back in history and (3) when the workbench unloads. [Here's how you can do 3](https://github.com/microsoft/vscode/blob/46860a105b2f2e115970ad0d54f5a39d5add9cc5/src/vs/platform/storage/common/storage.ts#L47-L47).

> do you agree this would be the expectation?

Absolutely! You can always change the `HistoryNavigator` itself to support your use case. How about adding a `remove()` method? That way you can remove something at will.

**Comment by meganrogge** [comment]
@joaomoreno this is working now except:

1. The ContextKey so that the cursor must be either on the first line (for up arrow) or last line (for down arrow) of the input box. I'm not sure how to do this despite looking at the links you sent me.
2. Regarding save on window reload, is the intended functionality that the state is preserved only when I've typed something and haven't yet committed it? or should it also preserve where you are in the history navigation? if the former, that is working. if the latter, then I will need to somehow store the actual `historyNavigator` itself since otherwise its state gets reset in the constructor when it's defined with the stored history values on window reload. `this.historyNavigator = new HistoryNavigator(JSON.parse(history), 50);`

**Comment by joaomoreno** [comment]
@meganrogge 

1. Yeah this isn't so clear to grasp at first. I've added a commit which sets these context keys. All you have to do now it hook those context keys into the `when` clauses of the keybindings.
2. When the workbench comes up, the user should always be pointing to the last element in the history. Right now I don't really see that happening, following these steps: (1) type something, (2) press Up to see the previous message in history and (3) reload -  the value I typed is gone. I expected to see the value I typed in step (1), as it is the last item in history.

A few other issues I see:

- (1) Open the workbench, box is empty, (2) press Up to see the last message, (3) press Down to go back to empty -> :bug: it doesn't
- (2) Open the workbench, box is empty, (2) type anything, (3) press Up -> 🐛 nothing happens, it only works if I press Up twice

**Comment by meganrogge** [comment]
@joaomoreno 
> * (1) Open the workbench, box is empty, (2) press Up to see the last message, (3) press Down to go back to empty -> 🐛 it doesn't*

I thought we had agreed that we want the SCM's behavior to mirror that of the search editor. If I have an empty search bar, when I go up and come back down, it remains on the most recently searched value as opposed to an empty box. 
recording (6)



> * (2) Open the workbench, box is empty, (2) type anything, (3) press Up -> 🐛 nothing happens, it only works if I press Up twice

Weird - I will investigate this. Maybe it's storing a duplicate value.

> Idea: why not allowing the history to store the empty string as a value? Raising that restriction will let you fix issue (1) above and will let you just store strings, instead of storing strings and booleans in history.

Yes that seems like a good idea.

**Comment by joaomoreno** [comment]
@meganrogge First off: this was a **hard** work item. Even I failed to grasp the complexity of the issue. As I reviewed it this time I decided to go ahead and finalize it by making some changes. You did a good job, but ultimately were hitting issues of the current HistoryNavigator. Let me explain.

The current HistoryNavigator is weird. It lets the navigation go out of its bound, which is really nasty and misaligns with the fact that we want to store the current editing value as the last of the history. This is the reason why you couldn't simply store strings. It was also the source of bugs like having to press `Up` twice or going beyond the history array bounds accidentally. I took the time to write a new HistoryNavigator2 which never lets the cursor go outside a history. This lets us just store strings, while keeping our code much smaller and easy to reason about. Please give it a look, try it out and let me know what you think.

The caveat with HistoryNavigator2 is that it doesn't remove duplicates, since I didn't think that was too important. But feel free to add that capability to HistoryNavigator2 if you think it is, should be pretty easy.

Some notes on the work:

- There were both instances of [bad formatting](https://github.com/microsoft/vscode/pull/107619/commits/526af22a5b37baea33c07839e3ba3d260f587dd3#diff-47107f7972200f9a87a45eac2bf289b735e3d59e81429d9407447ca966b8c625R65) and [eslint errors like using `==` instead of `===`](https://github.com/microsoft/vscode/commit/78b8a4f6b5159725bf416dd0494d3c2f1dba23d1#diff-0040e0ad33ad9fad03ce48fa1b94486305a362fc2d6ed0401aa1bc270369ff77R103). We currently have a precommit rule which prevents you from committing mistakes like this. How were you able to create these commits?
- It's usually good practice to [wrap `JSON.parse` in a `try catch`](https://github.com/microsoft/vscode/pull/107619/files#diff-0040e0ad33ad9fad03ce48fa1b94486305a362fc2d6ed0401aa1bc270369ff77R79-R83)

As you test this, remember to try it in a new workspace, since the storage in your test workspace contains the array of tuples, so the code will break, since it expects an array of strings.

**Review by joaomoreno — APPROVED** [review]

**Comment by meganrogge** [comment]
> We currently have a precommit rule which prevents you from committing mistakes like this. How were you able to create these commits?

probably by using the `--no-verify` flag on occasion. I won't do that going forward.

**Comment by meganrogge** [comment]
@joaomoreno it's now working as expected. I'll wait for your final sign-off to merge it.

**Review by joaomoreno — APPROVED** [review]
Your fix wasn't quite there yet. That change broke the _save on reload_ functionality. The right fix here would be to call `replaceLast` on the history only if the user is currently typing their message (`isAtEnd() === true`). My bad! 🙈  I've pushed those changes.

I've tested it more and feels great, what do you think? Merging this in! 💪 Further fixes can land in `master` directly, if necessary.

fireworks

**Pull request merged**
